sotodlib/io/bookbinder.py

```diff
@@ -583,21 +586,36 @@ def flush(self, flush_time=None):
 
         smurf_data, smurf_bias, smurf_primary = self.smbundle.rebundle(flush_time)
 
-        # Write signal data to frame
+        # Fill in missing samples and write data to frame
         f = core.G3Frame(core.G3FrameType.Scan)
-        f['signal'] = smurf_data
+        smurf_data, flag_smurfgap = self.fill_in_missing_samples(smurf_data, flush_time, return_flags=True)
+        f['signal'], data_excess = split_supertimestream(smurf_data, self.maxlength)
+        fx = {'data': data_excess}  # note input data use 'data' instead of 'signal'
         if smurf_bias is not None:
-            f['tes_biases'] = smurf_bias
+            smurf_bias = self.fill_in_missing_samples(smurf_bias, flush_time)
+            f['tes_biases'], fx['tes_biases'] = split_supertimestream(smurf_bias, self.maxlength)
         if smurf_primary is not None:
-            f['primary'] = smurf_primary
-
-        # Ancillary data object to be written to disk (co-sampled HK encoder data)
+            smurf_primary = self.fill_in_missing_samples(smurf_primary, flush_time)
```

Comment:
You're calling fill_in_missing_samples three different times here, but you really only need find the missing samples once since the timestamps are the same for these three arrays.

Reply:
If someone has time to optimize this, that would be appreciated! (This is also the reason why the same message gets printed 3 times, so it would be great if it didn’t do that)